test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java

```diff
@@ -0,0 +1,371 @@
+        log.info("Sealing stream {}", STREAM_NAME);
+        assertTrue(sealStreamStatus.get());
+        CompletableFuture<Boolean> deleteStreamStatus = controller.deleteStream(scope, stream);
+        log.info("Deleting stream {}", STREAM_NAME);
+        assertTrue(deleteStreamStatus.get());
+        CompletableFuture<Boolean> deleteScopeStatus = controller.deleteScope(scope);
+        log.info("Deleting scope {}", scope);
+        assertTrue(deleteScopeStatus.get());
+    }
+
+    void createScopeAndStream(String scope, String stream, StreamConfiguration config, URI uri) {
+
+        try (StreamManager streamManager = new StreamManagerImpl(uri)) {
+            Boolean createScopeStatus = streamManager.createScope(scope);
+            log.info("Creating scope with scope name {}", scope);
+            log.debug("Create scope status {}", createScopeStatus);
+            Boolean createStreamStatus = streamManager.createStream(scope, stream, config);
+            log.debug("Create stream status {}", createStreamStatus);
+        }
+    }
+
+    void createWriters(ClientFactory clientFactory, final int writers, String scope, String stream) {
+        log.info("Client factory details {}", clientFactory.toString());
+        log.info("Creating {} writers", writers);
+        writerList = new ArrayList<>(writers);
+        log.info("Writers writing in the scope {}", scope);
+        CompletableFuture.runAsync(() -> {
+                    for (int i = 0; i < writers; i++) {
+                        log.info("Starting writer{}", i);
+                        final EventStreamWriter<Long> tmpWriter = clientFactory.createEventWriter(stream,
+                                new JavaSerializer<Long>(),
+                                EventWriterConfig.builder().retryAttempts(10).build());
+                        writerList.add(tmpWriter);
+                        writerFutureList.add(startWriting(tmpWriter));
+                    }
+                });
+    }
+
+    void createReaders(ClientFactory clientFactory, String readerGroupName, String scope,
+                                 ReaderGroupManager readerGroupManager, String stream, final int readers) {
+        log.info("Creating Reader group: {}, with readergroup manager using scope: {}", readerGroupName, scope);
+        readerGroupManager.createReaderGroup(readerGroupName, ReaderGroupConfig.builder().startingTime(0).build(),
+                Collections.singleton(stream));
+        log.info("Reader group name: {}, Reader group scope: {}, Online readers: {}",
+                readerGroupManager.getReaderGroup(readerGroupName).getGroupName(), readerGroupManager
+                        .getReaderGroup(readerGroupName).getScope(), readerGroupManager
+                        .getReaderGroup(readerGroupName).getOnlineReaders());
+        log.info("Creating {} readers", readers);
+        readerList = new ArrayList<>(readers);
```

Comment:
Per my earlier comment: readerList should be a return value not a member variable.